File: ds-worker/src/main/java/com/example/dsworker/config/CaffeineConfig.java
package com.example.dsworker.config;

import com.github.benmanes.caffeine.cache.Cache;
import com.github.benmanes.caffeine.cache.Caffeine;
import com.github.benmanes.caffeine.cache.RemovalListener;
import com.zaxxer.hikari.HikariDataSource;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import java.util.concurrent.TimeUnit;

@Configuration
public class CaffeineConfig {

    /**
     * 数据源的缓存
     * @return
     */
    @Bean
    public Cache<String, HikariDataSource> dataSourceCache() {
        return Caffeine.newBuilder()
                .expireAfterWrite(60, TimeUnit.MINUTES)  // 一小时后过期
                .initialCapacity(30)    // 初始容量 30
                .maximumSize(200)       // 最大容量 200
                .removalListener((RemovalListener<String, HikariDataSource>) (key, dataSource, removalCause) -> {
                    if (dataSource != null) {
                        dataSource.close();
                    }
                })
                .build();
    }
}

File: api-factory/af-security-gateway/src/main/kotlin/org/inet/aet/afsecuritygateway/config/CaffeineConfig.kt
package org.inet.aet.afsecuritygateway.config

import com.github.benmanes.caffeine.cache.Cache
import com.github.benmanes.caffeine.cache.Caffeine
import com.github.benmanes.caffeine.cache.RemovalCause
import org.inet.aet.afsecuritygateway.entity.RouteLocal
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import java.util.concurrent.TimeUnit

@Configuration
class CaffeineConfig {

    /**
     * 缓存用户的权限信息
     */
    @Bean
    fun afPermCache(): Cache<String, List<String>> {
        return Caffeine.newBuilder()
            .expireAfterWrite(1, TimeUnit.MINUTES)
            .initialCapacity(30)
            .maximumSize(100)
            .build()
    }

    /**
     * 缓存 API 加密的信息
     */
    @Bean
    fun apiEncryptCache(): Cache<String, RouteLocal> {
        return Caffeine.newBuilder()
            .expireAfterWrite(1, TimeUnit.HOURS)
            .initialCapacity(30)
            .maximumSize(100)
            .removalListener { _: Any?, value: Any?, _: RemovalCause? -> (value as RouteLocal).curatorCache.close() }
            .build()
    }
}

